### DISCUSSION TITLE
Delayed transitions are cancelled when transitioning in parallel state

### DISCUSSION THREAD

**Discussion by rlaffers** [statelyai/xstate#2523] [General]
**Description**
Delayed transitions are never executed if an event triggered a transition in a parallel state.

```javascript
const machine = createMachine({
  id: "root",
  type: "parallel",
  states: {
    Operation: {
      initial: "Waiting",
      states: {
        Waiting: {
          after: {
            1000: "Fetching"
          },
          on: {
            // if this transition is triggered, the delay timeout
            // gets cancelled (bug?)
            TOGGLE_MODE: "#root.Mode.Demo"
          }
        },
        Fetching: {}
      }
    },
    Mode: {
      initial: "Normal",
      states: {
        Normal: {},
        Demo: {}
      }
    }
  }
})
```

**Expected Result**
In the linked CodeSandbox, the expected events logged in the console are:
```
{type: "xstate.init"}
{type: "xstate.after(1000)#root.Operation.Waiting"}
```
Regardless of whether the "TOGGLE_MODE" event was fired, the machine remains in the Operation.Waiting state,
therefore the delayed transition to Operation.Fetching should occur after 1 second.

**Actual Result**
However, if "TOGGLE_MODE" event is fired while in the Operation.Waiting state, the delayed transition to Operation.Fetching never occurs. These are the logged events:
```
{type: "xstate.init"}
{type: "TOGGLE_MODE"}
```

**Reproduction**
https://codesandbox.io/s/modest-ritchie-4ho6i?file=/src/index.js

**Additional Info**
Adding `internal: true` on the TOGGLE_MODE transition is a partial workaround (usable in my particular case, not generally).
```javascript
TOGGLE_MODE: {
  target: '#root.Mode.Demo',
  internal: true, // now the interpreter will not forget about the delayed transition
},
```
However, setting `internal: false`, which is the default behavior when no `internal` property is present, causes the interpreter to ignore the delayed transition. It seems that external transition in a parallel state causes the interpreter incorrectly cancel the timeout on delayed transition, even though the machine remains in the same state ("Waiting").

**Comment by Andarist** [comment]
You can check out this discussion: https://github.com/statelyai/xstate/discussions/1829 to learn why this happens. You can also let us know what do you think about the problem, proposed changes, etc

**Reply by rlaffers** [reply]
Thanks for the linked discussion. As I understand it, the debate about whether transitions should be auto/internal/external by default is separate from this problem: knowing that the transition is external by default does not explain why the interpreter does not restart the delayed timer upon re-entering the Waiting state. 
To me this seems like a bug because the expectation is that it should be impossible for a machine to remain indefinitely in a state with a declared delayed transition. The whole idea of the "Waiting" state above is that it is transient and the machine will transition out of it after the delay has passed. This should happen regardless of whether some event is handled by an internal or external transition. Note, that after the TOGGLE_MODE event, the machine still *is* in the Operation.Waiting state, but no delayed transition is going to occur anymore.
